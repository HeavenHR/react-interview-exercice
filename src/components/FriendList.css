:local(.friendList) {
  padding-left: 0;
  margin-bottom: 0;
}

:local(.btnMargin) {
  margin: 0 3px;
}

:local(.currentPage) {
  background: cornflowerblue;
  color: white;
}

.activeButton {
  background: cornflowerblue;
  color: white;
}
