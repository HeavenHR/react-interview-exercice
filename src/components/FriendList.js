import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './FriendList.css';
import FriendListItem from './FriendListItem';
import { ITEMS_PER_PAGE } from '../constants/PagingSettings';

class FriendList extends Component {
  render () {
    const { itemsPerPage } = this.state;
    return (
      <div className={styles.friendListWrapper}>
        <div className={styles.friendListOptions}>
          Friends per page:
          <select
            value={itemsPerPage}
            onChange={e => this.setState({ itemsPerPage: e.target.value, page: 1 })}
            className={styles.friendListOptions__select}>
            {
              ITEMS_PER_PAGE.map(option => {
                return <option
                          key={option}
                          value={option}
                          className={styles.friendListOptions__option}>
                          {option}
                        </option>
              })
            }
          </select>
        </div>
        <ul className={styles.friendList}>
          {
            this.getPageItems().map((friend, index) => {
              return (
                <FriendListItem
                  key={index}
                  id={index}
                  name={friend.name}
                  starred={friend.starred}
                  {...this.props.actions} />
              );
            })
          }
        </ul>
        <div className={styles.friendListPaging}>
          <button
            onClick={() => this.goPrevPage()}
            disabled={this.isFirstPage()}
            className='btn btn-default btn-xs'>
            <i className="fa fa-caret-left"></i>
          </button>
          <button
            onClick={() => this.goNextPage()}
            disabled={this.isLastPage()}
            className='btn btn-default btn-xs'>
            <i className="fa fa-caret-right"></i>
          </button>
        </div>
      </div>
    );
  }

  constructor (props, context) {
    super(props, context);
    this.state = {
      page: 1,
      itemsPerPage: 2
    };
  }

  getPageItems () {
    const { page, itemsPerPage } = this.state;
    const { friends } = this.props;

    if(friends.length <= itemsPerPage) {
      return friends
    }

    const firstPageItem = (page * itemsPerPage) - itemsPerPage;
    return [...friends].splice(firstPageItem, itemsPerPage)
  }

  goPrevPage () {
    this.setState({ page: this.state.page - 1 })
  }

  goNextPage () {
    this.setState({ page: this.state.page + 1 })
  }

  isFirstPage () {
    return this.state.page === 1
  }

  isLastPage () {
    const { page, itemsPerPage } = this.state;
    const { friends } = this.props;
    return friends.length <= page * itemsPerPage
  }
}

FriendList.propTypes = {
  friends: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired
};

export default FriendList;
