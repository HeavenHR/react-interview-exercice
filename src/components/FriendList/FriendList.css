:local(.friendList) {
  padding-left: 0;
  margin-bottom: 0;
  min-height: 200px;
}
