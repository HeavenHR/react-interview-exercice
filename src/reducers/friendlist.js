import {random} from 'lodash'
import * as types from '../constants/ActionTypes';

const initialState = {
  friendsById: [
    {
      id:random(10000,1000000,false),
      name: 'Theodore Roosevelt',
      gender: 'male',
      starred: true
    },
    {
      id:random(10000,1000000,false),
      name: 'Abraham Lincoln',
      gender: 'male',
      starred: false
    },
    {
      id:random(10000,1000000,false),
      name: 'George Washington',
      gender: 'male',
      starred: false
    }
  ]
};

export default function friends(state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      return {
        ...state,
        friendsById: [
          ...state.friendsById,
          {
            name: action.friend.name,
            gender: action.friend.gender,
            id:random(10000,1000000,false),
          }
        ],
      };
    case types.DELETE_FRIEND:
      return {
        ...state,
        friendsById: state.friendsById.filter((item) => item.id !== action.id)
      };
    case types.STAR_FRIEND:
      let friends = [...state.friendsById];
      let friend = friends.find((item) => item.id === action.id);
      friend.starred = !friend.starred;
      return {
        ...state,
        friendsById: friends
      };

    default:
      return state;
  }
}
