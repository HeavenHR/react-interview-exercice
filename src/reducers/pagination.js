import { UPDATE_PAGINATION } from '../constants/ActionTypes';

export const initialState = {
  pageSize: 2,
  currentPage: 1,
  pagingRange: 3
};

export const pagination = (state = initialState, action) => {
  const { type, currentPage } = action;
  switch (type) {
    case UPDATE_PAGINATION: {
      return {
        ...state,
        currentPage
      };
    }

    default:
      return state;
  }
};
