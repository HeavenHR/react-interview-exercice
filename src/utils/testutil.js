import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
