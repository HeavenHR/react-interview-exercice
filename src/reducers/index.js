import { combineReducers } from 'redux';
import { default as friendlist } from './friendlist';
import { pagination } from './pagination';

export default combineReducers({ friendlist, pagination });
