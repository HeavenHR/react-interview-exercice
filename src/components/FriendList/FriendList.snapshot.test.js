import React from 'react';

import FriendList from './FriendList';
import { createMountedComponent } from '../../utils/testutil';
import { derivePageItems } from '../../utils/pagination';
import { initialState  } from '../../reducers/friendlist';

describe('<FriendList /> should', () => {
  const pageSize = 2;
  const currentList = derivePageItems(initialState.friendsById, 1, pageSize);

  it('render correctly according to props', () => {
    const mountedFriendList = createMountedComponent({
      children: <FriendList friends={currentList} actions={localStorage.actions} />
    });

    const listItems = mountedFriendList.find('li.list-group-item');

    expect(mountedFriendList.htmlRender()).toMatchSnapshot();
    expect(listItems.length).toEqual(2);
  });
});
