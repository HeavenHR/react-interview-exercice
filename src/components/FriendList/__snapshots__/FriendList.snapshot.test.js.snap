exports[`<FriendList /> should render correctly according to props 1`] = `
"<ul class=\"list-group undefined\">
  <li class=\"list-group-item undefined\">
    <div>
      <div><span>Theodore Roosevelt</span>
      </div>
      <div><i class=\"fa fa-male\"></i><small>male</small>
      </div>
      <div><small>xx friends in common</small>
      </div>
    </div>
    <div>
      <button class=\"btn btn-default undefined\"><i class=\"fa fa-star\"></i>
      </button>
      <button class=\"btn btn-default undefined\"><i class=\"fa fa-trash\"></i>
      </button>
    </div>
  </li>
  <li class=\"list-group-item undefined\">
    <div>
      <div><span>Abraham Lincoln</span>
      </div>
      <div><i class=\"fa fa-male\"></i><small>male</small>
      </div>
      <div><small>xx friends in common</small>
      </div>
    </div>
    <div>
      <button class=\"btn btn-default undefined\"><i class=\"fa fa-star-o\"></i>
      </button>
      <button class=\"btn btn-default undefined\"><i class=\"fa fa-trash\"></i>
      </button>
    </div>
  </li>
</ul>"
`;
